File: 4949.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String args[]) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        while(true) {
            String s = br.readLine();
            
            if(s.equals(".")) {
                return;
            }
            
            Stack<Character> stack = new Stack<>();
            
            for(int i = 0; i < s.length(); i++) {
                if(s.charAt(i) == '(') {
                    stack.push(')');
                }
                
                else if(s.charAt(i) == '[') {
                    stack.push(']');
                }
                
                else if(s.charAt(i) == ')' || s.charAt(i) == ']') {
                    if(!stack.isEmpty() && stack.peek() == s.charAt(i)) {
                        stack.pop();
                    }
                    
                    else {
                        stack.push('x');
                    }
                }
            }
            
            System.out.println(stack.isEmpty() ? "yes" : "no");
        }
    }
}

File: 4949.cpp
#include<bits/stdc++.h>
using namespace std;

string s;

int main() {
	while(true) {
		getline(cin, s);
		
		if(s.size() == 1 && s[0] == '.') {
			break;
		}
		
		bool chk = true;
		stack<char> st;
		
		for(int i = 0; i < s.size(); i++) {
			if(s[i] == '('|| s[i] == '[') {
				st.push(s[i]);
			}
			
			else if(s[i] == ')') {
				if(st.size() > 0 && st.top() == '(') {
					st.pop();
				}
				
				else {
					chk = false;
					
					break;
				}
			}
			
			else if(s[i] == ']') {
				if(st.size() > 0 && st.top() == '[') {
					st.pop();
				}
				
				else {
					chk = false;
					
					break;
				}
			}
		}
		
		if(st.empty() && chk) {
			cout << "yes" << endl;
		}
		
		else {
			cout << "no" << endl;
		}
	}
}